resolve fastify-mysql promise causing tap timeouts #33

jerrythomas opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Tip: `npm run bench` to compare branches interactively.

Contributors guide: https://github.com/fastify/fastify/blob/master/CONTRIBUTING.md
-->
## Issue

While developing an app using `fastify-mysql`, I observed that my tests were timing out. If I disable `promise`, the same tests were not timing out. I tried but could not replicate this using a standalone test. It seems like `fastify` is waiting for MySQL to close the connection.

As mentioned in [issue #33] (https://github.com/fastify/fastify-mysql/issues/33) changing the code for the onClose event also resolved this issue for me. 

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

```
> fastify-mysql@1.0.0 test /Users/Jerry/Code/issues/fastify-mysql
> standard && tap test/*.test.js

test/connection.test.js ............................. 38/38
test/initialization.test.js ......................... 19/19
test/pool.promise.test.js ........................... 13/13
test/pool.test.js ................................... 35/35
total ............................................. 105/105

  105 passing (2s)

  ok
```

mcollina commented on index.js:
```diff
@@ -17,7 +17,13 @@ function fastifyMysql (fastify, options, next) {
-    fastify.addHook('onClose', (fastify, done) => client.end(done))
+    if (usePromise) {
+      fastify.addHook('onClose', async (fastify, done) => {
+        await client.end(); done()
```
I don't think the call to `done()` is needed. Could you remove?

jerrythomas replied:
I can't figure out how to test this. Any suggestions on an approach and I will make the changes.

Eomm replied:
The problem here is that we should use the sync or async approach based on the `usePromise` option:

if it is true the hook should be:

```
fastify.addHook('onClose', async (fastify) => {
  await client.end();
})
```

if it is falsy the hook should manage the `done`

We can merge those in:

```
fastify.addHook('onClose', (fastify, done) => {
  const end = client.end()
  if (end && typeof end.then === 'function' ){
    end.then(done, done)
  } else {
    done()
  }
})
```

jerrythomas commented:
Added a couple of unit tests and modified the close hook. Using the `async await` on `pool.end()` works in both promise and callback options. The tests do not complete if the `onClose` callbacks are not present.

jsumners commented on index.js:
```diff
@@ -17,7 +17,9 @@ function fastifyMysql (fastify, options, next) {
     const client = connectionType !== 'connection' ? db.pool : db.connection
 
-    fastify.addHook('onClose', (fastify, done) => client.end(done))
+    fastify.addHook('onClose', async (fastify, done) => {
+      await client.end(); done()
+    })
```
I really unclear as to how this change is any different from the original.

jerrythomas replied:
All I understood is that when `mysql` promise is true `client.end(done)` callback function is not valid and `client.end()` is an async function. If I revert it to the old one, the new tests fail. 

```text
test/closing.test.js .................................. 6/7
  promise connection
  not ok test unfinished
    stack: |
      Object.<anonymous> (test/closing.test.js:29:1)
      Generator.next (<anonymous>)
      internal/main/run_main_module.js:17:47
    test: promise connection
    at:
      line: 29
      column: 1
      file: test/closing.test.js
      function: Object.<anonymous>
    source: |
      test('promise connection', (t) => {

test/connection.test.js ............................. 38/38
test/initialization.test.js ......................... 19/19
test/pool.promise.test.js ........................... 13/13
test/pool.test.js ................................... 35/35
total ............................................. 111/112
  

  111 passing (3s)
  1 failing
```

and with the new code
```text
> fastify-mysql@1.0.0 test /Users/Jerry/Code/issues/fastify-mysql
> standard && tap test/*.test.js

test/closing.test.js .................................. 7/7
test/connection.test.js ............................. 38/38
test/initialization.test.js ......................... 19/19
test/pool.promise.test.js ........................... 13/13
test/pool.test.js ................................... 35/35
total ............................................. 112/112

  112 passing (3s)

  ok
```

Eomm replied:
When we use `require('mysql2/promise')`, `client.end()` return every time a Promise, even if a callback is provided

https://github.com/sidorares/node-mysql2/blob/master/promise.js#L64

jsumners replied:
I think it would be better to use the `usePromise` boolean to determine if it should be:

```js
client.end(done)
```

or

```js
client.end().then(done).catch(done)
```

jerrythomas commented:
@jsumners changed as per your suggestion.

jerrythomas commented:
Should there be a version change and release to npm?

Eomm commented:
Already Released as 1.0.1 👍🏼
